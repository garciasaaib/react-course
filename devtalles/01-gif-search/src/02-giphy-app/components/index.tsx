export { AddCategory } from './AddCategory';
export { GifGrid } from './GifGrid';
export { GifItem } from './GifItem';