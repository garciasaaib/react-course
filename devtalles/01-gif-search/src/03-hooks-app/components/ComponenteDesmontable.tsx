import { useEffect, useState } from "react"

export const ComponenteDesmontable = () => {
  const [coords, setCoords] = useState({x:0, y:0})
  // reglas de montaje y desmontaje
  useEffect(() => {
    console.log('Message Mounted')
    return () => {
      console.log('Message Unmounted')
    }
  }, [])

  // efecto encadenado por el movimiento del mouse
  useEffect(() => {
    // creamos la funcion
    const onMouseMove = ({x, y}: MouseEvent) => {
      const coords = {x,y}
      setCoords(coords)
    }

    // agregamos esta funcion por referencia cuando se monta
    window.addEventListener('mousemove', onMouseMove)

    // quitamos la funcion del evento por referencia
    return () => {
      window.removeEventListener('mousemove',onMouseMove)
    }
  },[])
  return (
    <>
      <h3>Usuario ya existe</h3>
      <p>x: {coords.x}</p>
      <p>y: {coords.y}</p>
    </>
  )
}
