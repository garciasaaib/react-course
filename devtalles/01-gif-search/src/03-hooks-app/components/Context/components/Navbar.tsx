import { Link, useLocation } from "react-router";

export const Navbar = () => {
  const {pathname} = useLocation()
  const onIsActive = (thisPath: string) => {
    return pathname === thisPath ? " underline" : "";

  }
	return (
		<nav className="flex gap-1">
			<Link className={`hover:text-white${onIsActive('/')}`} to="/">
				Home
			</Link>
			<Link className={`hover:text-white${onIsActive('/about')}`} to="/about">
				About
			</Link>
			<Link className={`hover:text-white${onIsActive('/login')}`} to="/login">
				Login
			</Link>
		</nav>
	);
};
