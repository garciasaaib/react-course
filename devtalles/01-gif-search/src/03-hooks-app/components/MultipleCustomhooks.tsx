export const MultipleCustomhooks = () => {
  return (
    <div>MultipleCustomhooks</div>
  )
}