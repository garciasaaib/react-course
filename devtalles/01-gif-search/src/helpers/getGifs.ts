import { GiphyResponseSuccess } from "../api/giphy"


export type Gif = {
  id: string
  title: string
  url: string
}

export type GifFetchResult =
  | { success: true; gifs: Gif[], error: false }
  | { success: false; gifs: false, error: string }

const api_key = "Go84so0o7WglUIx6lqnHAbYWvgdwEdOR"
const host = "https://api.giphy.com/v1/gifs/search"
const limit = "10"

export const getGifs = async (category: string): Promise<GifFetchResult>  => {
  const url = `${host}?api_key=${api_key}&q=${encodeURI(category)}&limit=${limit}`
  try {
    const resp = await fetch(url)
    if (!resp.ok) {
      throw new Error(`HTTP error! status: ${resp.status}`)
    }
    const { data }: GiphyResponseSuccess = await resp.json()
    const gifs = data.map((gif) => ({
      id: gif.id,
      title: gif.title,
      url: gif.images?.downsized_medium.url
    }))
    return { success: true, gifs, error: false }
  } catch (error: unknown) {
    return { success: false, gifs: false, error: (error as Error).message }
  }
}