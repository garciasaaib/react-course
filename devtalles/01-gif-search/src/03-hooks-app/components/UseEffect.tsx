import { ChangeEvent, useEffect, useState } from "react";
import { ComponenteDesmontable } from "./ComponenteDesmontable";

export const UseEffect = () => {
  const [state, setState] = useState({
    username: 'Adrian',
    email: 'adrian@gmail.com',
  })
  const {username, email} = state
  const onInputChange = ({target}: ChangeEvent<HTMLInputElement>) => {
    const {value, name} = target
    setState({...state, [name]: value})
  }
  useEffect(() => {
    // body: all that this does


    // return () => {
      // clean up: cleanning for unmount component
    //}
  }, []) // dependencies: watching variables
	return (
		<>
			<h1>SimpleForm</h1>
			<hr />
			<form>
				<input onChange={onInputChange} value={username} type="text" placeholder="Username" name="username" />
				<input onChange={onInputChange} value={email} type="email" placeholder="Email" name="email" />
			</form>
      {/* El monte y desmonte se genera desde el compoonente padre */}
      { username === 'Adrian' && <ComponenteDesmontable /> }
		</>
	);
};
