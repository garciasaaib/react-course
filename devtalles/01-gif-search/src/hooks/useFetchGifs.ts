import { useEffect, useState } from "react"
import { getGifs, Gif } from "../helpers/getGifs"

export const useFetchGifs = (category: string) => {
  const [images, setImages] = useState<Gif[]>([])
  const [loading, setLoading] = useState(true)
  const handleSetImages = async () => {
    setLoading(true)
    try {
      const {gifs} = await getGifs(category)
      // eslint-disable-next-line @typescript-eslint/no-unused-expressions
      gifs && setImages(gifs)

    }
    catch (error: unknown) {
      console.error((error as Error).message)
    }
    finally {
      setLoading(false)
    }
  }
  useEffect(() => {
    handleSetImages()

    // cleanup: return a function that will be executed when the component is unmounted
    // and is used to cancel any pending request and avoid memory leaks, and clean events
  }, [])
  return {
    gifs: images,
    loading
  }
}
