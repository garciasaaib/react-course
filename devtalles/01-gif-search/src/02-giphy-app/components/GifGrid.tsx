import { GifItem } from "./";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }: { category: string }) => {
	const { gifs, loading } = useFetchGifs(category);

	return (
		<>
			<h3>{category}</h3>
			{loading && <h2>Cargando...</h2>}

			<div className="card-grid">
				{gifs.map((image) => (
					<GifItem key={image.id} {...image} />
				))}
			</div>
		</>
	);
};
