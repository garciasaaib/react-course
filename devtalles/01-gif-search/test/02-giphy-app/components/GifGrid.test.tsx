import React from "react";
import { render, screen } from "@testing-library/react";
import { GifGrid } from "../../../src/02-giphy-app/components/GifGrid";
import { useFetchGifs } from "../../../src/02-giphy-app/hooks/useFetchGifs";
import { Gif } from "../../../src/02-giphy-app/helpers/getGifs";

const category = "Saitama";
// con esto, mockeamos useFetchGifs
jest.mock("../../../src/02-giphy-app/hooks/useFetchGifs");

// le decimos que el hook useFetchGifs es una funcion mockeada tipo useFetchGifs
// y que su valor de retorno por defecto es el designado
(useFetchGifs as jest.MockedFunction<typeof useFetchGifs>).mockReturnValue({
	gifs: [],
	loading: true,
});

describe("<GifGrid />", () => {
	test("Debe crear el loading", () => {
		render(<GifGrid category={category} />);
		// evaluamos que diga cargando
		expect(screen.getByText("Cargando..."));
		// evaluar que diga la categoria en algun lado
		expect(screen.getByText("Cargando..."));
	});

	test("Debe mostrar un listado de items", () => {
		render(<GifGrid category={category} />);
		// evaluamos que diga cargando
		expect(screen.getByText("Cargando..."));
		// evaluar que diga la categoria en algun lado
		expect(screen.getByText("Cargando..."));
	});

	test("debe de mostrar items cuando se cargan las imagenes useFetchGifs", () => {
		// creamos el mock de la lista de gifs
		const gifs: Gif[] = [
			{
				id: "1",
				title: "Saitama",
				url: "http://localhost/saitama.jpg",
			},
			{
				id: "2",
				title: "goku",
				url: "http://localhost/goku.jpg",
			},
		];

    // insertamos los gifs en el mock para esta prueba
    (useFetchGifs as jest.MockedFunction<typeof useFetchGifs>).mockReturnValue({
      gifs,
      loading: false,
    });
		// renderizamos el componente
		render(<GifGrid category={category} />);
    // evalua si el conteo de imagenes en el grid son dos
    expect(screen.getAllByRole('img').length).toBe(2)
	});
});
