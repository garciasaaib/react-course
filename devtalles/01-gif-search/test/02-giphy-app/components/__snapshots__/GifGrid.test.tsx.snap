// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifGrid /> Debe crear el loading 1`] = `
<div>
  <h3>
    Saitama
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
